File: tasks-core-jvm/src/main/java/org/funfix/tasks/DelayedFun.java
package org.funfix.tasks;

import org.jspecify.annotations.NullMarked;

/**
 * Represents a delayed computation (a thunk).
 * <p>
 * These functions are allowed to trigger side effects, with
 * blocking I/O and to throw exceptions.
 *
 * @see AsyncFun
 */
@NullMarked
@FunctionalInterface
public interface DelayedFun<T, E extends Exception> {
    T invoke() throws E;
}

File: tasks-core/src/jvmMain/kotlin/org/funfix/tasks/DelayedFun.kt
package org.funfix.tasks

import org.jetbrains.annotations.Blocking

/**
 * Represents a delayed computation (a thunk).
 *
 * These functions are allowed to trigger side effects, with
 * blocking I/O and to throw exceptions.
 *
 * @see [AsyncFun]
 */
public fun interface DelayedFun<out T> {
    @Blocking
    @Throws(Exception::class)
    public operator fun invoke(): T
}
